### ISSUE TITLE
automatic LaTeX rendering

### ISSUE THREAD

**Issue by goropikari** [korsbo/Latexify.jl#4]
Because I am a lazy person, I want to omit typing `display("text/latex", latexalign(f))` for LaTeX rendering.

I want LaTeX rendered output automatically as shown in the picture bellow.
![mathjax_auto](https://user-images.githubusercontent.com/21122112/33159064-034bc0ca-d053-11e7-8e26-d1d667d8bdc9.png)

Do you have any plans to implement such feature?

**Comment by korsbo** [comment]
I am indeed considering this, but I have not yet found a pleasing solution. 

I don't wish to remove or obfuscate the ability to print out the latex code for some copy-pasting (this was ever so useful when I wrote the docs). And `display("text/latex", x )` errors if you run Julia from the REPL or the console. 


it would be easy to add some 
```julia
latex_align_display(x...) = display("text/latex", latexalign(x...))
# or even a short alias
lad(x...) = latex_align_display(x...)
```
but I'm a bit wary of cluttering the namespace. Not so much due to fear of naming conflicts, but just because it would be difficult to remember all the different functions.  

Rather than doing this, I have been playing with the idea of reserving the main function name `latexify` to a function which tries to infer the best way of presenting the latex object to the user. The current `latexify` function would then be renamed to something like `latexraw`. `latexify` would then either print out the string form, or display a rendered version, possibly depending on keyword arguments or by checking if display("text/latex", x ) would throw an error. 

So, in short: yes, I want such a feature, but I have not decided on how this is best done.

Comments, ideas, suggestions, and cool solutions would be received with gratitude.

**Comment by goropikari** [comment]
[Pretty printing article](https://docs.julialang.org/en/stable/manual/types/#Custom-pretty-printing-1) in julia document is useful for you. [SymPy.jl codes](https://github.com/JuliaPy/SymPy.jl/blob/e3fba6f2130a7f716f9e8d48d73aeed6712892fa/src/display.jl#L49-L74) are also useful.

You don't have to define new fuctions, although you should define new type.
I think the easiest way to implement such features is as follows.
- Define new subtype of AbstractString, for example, LaTeXString. (Although the type name conflicts [LaTeXStrings.jl](https://github.com/stevengj/LaTeXStrings.jl))
```julia
struct LaTeXString <: AbstractString
    s::String
end
```
- Change the type of output, String, to new defined  subtype.
```julia
return str --> return LaTeXString(str)
```

- Define pretty printing for new subtype.
```julia
show(io::IO, ::MIME"text/latex", s::LaTeXString) = print(io, "\$" * s.s * "\$")
```


Then we can get LaTeX rendered output automatically in jupyter and get LaTeX format command by `print(latexalign(f))`.
Above method doesn't affect REPL output, so there is no error.


If you accept defining new subtype, I think this is the easiest solution.

The following example are my simple modification and jupyter sample. I wrote it quickly and I haven't tested. Probably there are bugs. So please take it as an example.
https://github.com/goropikari/Latexify.jl/commit/6016fe1bda9116064f69826eb11a90b9155911ea
http://nbviewer.jupyter.org/gist/goropikari/4b74cb8b087a5ac3131b003ebda8c41b

Of course, some people always want to display LaTeX command instead of LaTeX rendered output. So if we can switch output types, it is more convenient.

**Comment by korsbo** [comment]
This is awesome! 

I noticed that LaTeXString already has something like this implemented. Can you think of a reason not to just use their type?

**Comment by korsbo** [comment]
Ok, I just tried to use the LaTeXStrings.jl type and it has some side-effects. But the general idea is great and must be implemented!

**Comment by goropikari** [comment]
I am looking forward to the next version release :smile:
